### PR TITLE
[31973] Allow incoming email to log as Work Package comment

### PR DESCRIPTION
**Pull request by Givo29**
ref: https://community.openproject.com/projects/openproject/work_packages/31973/activity

### CODE REVIEW

**Comment by b12f** [comment]
Hi @Givo29 and please excuse us for the really long wait. I see tests are missing for sure, but before you put in extra work let me just check with the team if this would close the mentioned work package appropriately.

**Comment by cleary** [comment]
@b12f thanks for the reply!

@Givo29 and I work together, I raised the original ticket in OpenProject and got Joel to do the work to meet the ticket requirements.

In so far as we are concerned, it meets the ticket requirements and can be closed.

**Comment by b12f** [comment]
Hi @cleary and thanks for the info! I'll try to get this running on pullpreview and get some feedback from our team.

File: app/models/mail_handler.rb [hunk 1]
```diff
@@ -126,6 +126,16 @@ def options=(value)
     options[:no_permission_check] = options[:no_permission_check].to_s == '1'
   end
 
+  def reply_to_wp_from_receiver_addresses
+    addresses = [email.to, email.cc].flatten.compact.uniq.map { |a| a.strip.downcase }
+    unless addresses.empty?
```
Comment (ulferts) [inline app/models/mail_handler.rb:131]: I know that it is better to be safe than sorry but I cannot see a way the `addresses` can be empty. How did the mail get to the code if it is?

File: app/models/mail_handler.rb [hunk 2]
```diff
@@ -126,6 +126,16 @@ def options=(value)
     options[:no_permission_check] = options[:no_permission_check].to_s == '1'
   end
 
+  def reply_to_wp_from_receiver_addresses
+    addresses = [email.to, email.cc].flatten.compact.uniq.map { |a| a.strip.downcase }
```
Comment (ulferts) [inline app/models/mail_handler.rb:130]: This line of code is added 3 times within the PR and should thus be extracted into a method.

File: app/models/mail_handler.rb [hunk 3]
```diff
@@ -126,6 +126,16 @@ def options=(value)
...
+    unless addresses.empty?
+      addressMatch = addresses.find { |addr| /\+(\d+)/ =~ addr }
+      unless addressMatch.nil?
+        receive_work_package_reply(addressMatch.match(/\d+/).to_s)
```
Comment (ulferts) [inline app/models/mail_handler.rb:134]: The addresses are parsed two times here which makes the code harder to understand. The better approach would be to parse the addresses for the matching `/\+(\d+)/` segment (ideally only on the emission email address) and have that returned by a method.

File: app/models/mail_handler.rb [hunk 4]
```diff
@@ -548,9 +561,26 @@ def update_work_package(work_package)
...
+        cc_addresses = ''
+      end
+      attributes.merge(
+        'custom_field_values' => custom_field_values_from_keywords(work_package),
+        'journal_notes' => "Subject: #{email.subject.to_s.chomp[0, 255]}
+          To: #{email.to.join(', ')}
+          CC: #{cc_addresses}
+          Date: #{email.date.strftime('%d/%m/%y')}
+          Body:
+          #{cleaned_up_text_body}")
```
Comment (ulferts) [inline app/models/mail_handler.rb:578]: The header is not i18ned. A french person might want to have e.g. `Body` translated.

File: app/models/mail_handler.rb [hunk 5]
```diff
@@ -548,9 +561,26 @@ def update_work_package(work_package)
...
+      unless email.cc.nil?
+        cc_addresses = email.cc.join(', ')
+      else
+        cc_addresses = ''
+      end
+      attributes.merge(
+        'custom_field_values' => custom_field_values_from_keywords(work_package),
+        'journal_notes' => "Subject: #{email.subject.to_s.chomp[0, 255]}
+          To: #{email.to.join(', ')}
+          CC: #{cc_addresses}
```
Comment (ulferts) [inline app/models/mail_handler.rb:575]: `email.cc&.join(', ')` will suffice here.

**Review by ulferts — CHANGES_REQUESTED** [review]
Hi @cleary, and @Givo29,

and once again apologies for getting back to you with this much delay.

I like the added functionality of being able to add to a work package from an e-mail client without having to have received a matching e-mail first (or modifying some headers by hand).

On the user level I have a couple of concerns/questions. Comments on the code level I have inlined:
* The change will lead to OpenProject parsing every mail address for `/+\d+/`. This runs the risk of false positives. If some other recipient were to have a matching pattern in her/his mail address and that one was ordered before the intended reference, the update would become derailed altogether. How about we only parse the mail address used by the OpenProject instance for receiving updates by e-mail. It will be in the list of cc/to anyway.
* The change will display the header attributes (Subject, Date, To, Cc) only for updates received with a +\d e-mail address. But updates received by responding to received updates (e.g. identified by the header) will not have those header attributes displayed. To me, either all or no updates received via e-mails should have this header. Other tools do not display such a header at all. We might thus want to add a setting to toggle this functionality. 

Additionally, tests are missing.

**Comment by cleary** [comment]
@ulferts thanks for the detailed review, we will spend some time on it over the next couple of weeks (in between the other competing work, hopefully)

**Comment by ulferts** [comment]
Closing old PRs

**Pull request closed without merging**